extensions/spark/kyuubi-spark-jvm-quake/src/main/scala/org/apache/spark/kyuubi/jvm/quake/KyuubiJVMQuake.scala

```diff
@@ -0,0 +1,146 @@
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.kyuubi.jvm.quake
+
+import java.io.File
+import java.lang.management.ManagementFactory
+import java.util.concurrent.{ScheduledExecutorService, TimeUnit}
+
+import scala.collection.JavaConverters._
+
+import KyuubiJVMQuake._
+import com.sun.management.HotSpotDiagnosticMXBean
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.kyuubi.jvm.quake.KyuubiJVMQuakeConf._
+import org.apache.spark.util.ThreadUtils
+
+class KyuubiJVMQuake(conf: SparkConf, heapDumpEnabled: Boolean) extends Logging {
+  private val appId = conf.get("spark.app.id", "app-id").stripPrefix("application_")
+  private val dumpThreshold = TimeUnit.SECONDS.toNanos(conf.get(JVM_QUAKE_DUMP_THRESHOLD))
+  private val killThreshold = TimeUnit.SECONDS.toNanos(conf.get(JVM_QUAKE_KILL_THRESHOLD))
+  private val exitCode = conf.get(JVM_QUAKE_EXIT_CODE)
+  private val checkInterval = TimeUnit.SECONDS.toMillis(conf.get(JVM_QUAKE_CHECK_INTERVAL))
+  private val runTimeWeight = conf.get(JVM_QUAKE_RUN_TIME_WEIGHT)
+  private[quake] val heapPath = s"${conf.get(JVM_QUAKE_HEAP_DUMP_PATH)}/$appId"
+
+  private[quake] var (lastExitTime, lastGCTime) = getLastGCInfo
+  private[quake] var bucket: Long = 0L
+  private[quake] var heapExist: Boolean = false
+
+  private val scheduler: ScheduledExecutorService =
+    ThreadUtils.newDaemonSingleThreadScheduledExecutor("jvm-quake")
+
+  private[quake] def run(): Unit = {
+    val (currentExitTime, currentGcTime) = getLastGCInfo
+    if (currentExitTime != lastExitTime) {
+      val gcTime = currentGcTime - lastGCTime
+      val runTime = currentExitTime - lastExitTime - gcTime
+
+      bucket = Math.max(0, bucket + gcTime - BigDecimal(runTime * runTimeWeight).toLong)
+
+      if (bucket > dumpThreshold) {
+        logError(s"JVM GC has reached the threshold!!!" +
+          s" (bucket: ${bucket / 1000000000}s, dumpThreshold: ${dumpThreshold / 1000000000}s)")
+        if (shouldDumpHeap) {
+          saveHeap(heapPath)
+          heapExist = true
+        } else if (bucket > killThreshold) {
+          System.exit(exitCode)
+        }
+      }
+
+      lastExitTime = currentExitTime
+      lastGCTime = currentGcTime
+    }
+  }
+
+  def start(): Unit = {
+    scheduler.scheduleAtFixedRate(
+      new Runnable() {
+        override def run(): Unit = {
+          KyuubiJVMQuake.this.run()
+        }
+      },
+      0,
+      checkInterval,
+      TimeUnit.MILLISECONDS)
+  }
+
+  def stop(): Unit = {
+    scheduler.shutdown()
+  }
+
+  private def shouldDumpHeap: Boolean = {
+    heapDumpEnabled && !heapExist
```

Comment:
If quake is positioned as a guardian, I don't think heapExist is needed, that is, there is no need to dump many times, as soon as the threshold is reached, we dump and kill, rather than dump many times. This feels a bit like analyzing heap memory, but its goal is not that